rbac/management/principal/it_service.py

```diff
@@ -0,0 +1,347 @@
+        elif len(usernames) > 0:
+            service_account_principals = service_account_principals.filter(username__in=usernames)
+
+        # Sort order which defaults to ascending.
+        sort_order_ascending = True
+        sort_order = options.get("sort_order")
+        if sort_order:
+            sort_order_ascending = sort_order == "asc"
+
+        asc_order_by_enabled = True
+        order_by = options.get("order_by")
+        if order_by:
+            asc_order_by_enabled = not order_by.startswith("-")
+
+        if sort_order_ascending and asc_order_by_enabled:
+            service_account_principals = service_account_principals.order_by("username")
+        else:
+            service_account_principals = service_account_principals.order_by("-username")
+
+        # We always set a default offset and a limit if the user doesn't specify them so it is safe to simply put the
+        # two parameters in the query to slice it.
+        offset = options.get("offset")
+        limit = options.get("limit")
+
+        service_account_principals[offset:limit]
+
+        # If we are in an ephemeral or test environment, we will take all the service accounts of the user that are
+        # stored in the database and generate a mocked response for them, simulating that IT has the corresponding
+        # service account to complement the information.
+        if settings.IT_BYPASS_IT_CALLS:
+            it_service_accounts = self._get_mock_service_accounts(
+                service_account_principals=service_account_principals
+            )
+
+        # Put the service accounts in a dict by for a quicker search.
+        sap_dict: dict[str, dict] = {}
+        for sap in service_account_principals:
+            sap_dict[sap.service_account_id] = sap
+
+        # Filter the incoming service accounts. Also, transform them to the payload we will
+        # be returning.
+        service_accounts: [dict] = []
+
+        # If the "username_only" parameter was set, we should only return that for the user.
+        username_only = options.get("username_only")
+        if username_only and username_only == "true":
+            for it_service_account in it_service_accounts:
+                try:
+                    sa_principal = sap_dict[it_service_account["clientID"]]
+                    service_accounts.append({"username": sa_principal.username})
+                # If we cannot find a requested service account to IT in the database, we simply
+                # skip them.
+                except KeyError:
+                    continue
+        else:
+            for it_service_account in it_service_accounts:
+                try:
+                    sa_principal = sap_dict[it_service_account["clientID"]]
+
+                    it_service_account["username"] = sa_principal.username
+
+                    service_accounts.append(it_service_account)
+                # If we cannot find a requested service account to IT in the database, we simply
```

Comment:
It might be a bit cleaner to make a single loop and move the `if username_only and username_only == "true"` check into the body of the for loop since both for loops are constructing the service_accounts list.

Also, it looks like this logic is repeated.  Have you thought about moving this to a function?

Reply:
Great suggestion, yes! I've refactored it to a single function. Thanks!